contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,773 @@
+        uint32(block.timestamp + 1), // start
+        SafeUintCast.toUint32(params.period), // period
+        priceRatio,
+        false, // expired,
+        globalRewardsRatio
+      ));
+
+    return totalPremiumInNXM;
+  }
+
+  function allocateCapacity(
+    BuyCoverParams memory params,
+    uint coverId,
+    IStakingPool _stakingPool,
+    uint amount
+  ) internal returns (uint coveredAmountInNXM, uint premiumInNXM, uint rewardsInNXM) {
+
+    Product memory product = _products[params.productId];
+    uint gracePeriod = _productTypes[product.productType].gracePeriodInDays * 1 days;
+
+
+    return _stakingPool.allocateStake(
+      CoverRequest(
+        coverId,
+        params.productId,
+        amount,
+        params.period,
+        gracePeriod,
+        globalCapacityRatio,
+        product.capacityReductionRatio,
+        globalRewardsRatio
+      )
+    );
+  }
+
+  function editCover(
+    uint coverId,
+    BuyCoverParams memory buyCoverParams,
+    PoolAllocationRequest[] memory poolAllocations
+  ) external payable onlyMember whenNotPaused {
+
+    CoverData memory cover = _coverData[coverId];
+    uint lastCoverSegmentIndex = _coverSegments[coverId].length - 1;
+    CoverSegment memory lastCoverSegment = coverSegments(coverId, lastCoverSegmentIndex);
+
+    require(ICoverNFT(coverNFT).isApprovedOrOwner(msg.sender, coverId), "Cover: Only owner or approved can edit");
+    require(lastCoverSegment.start + lastCoverSegment.period > block.timestamp, "Cover: cover expired");
+    require(buyCoverParams.period < MAX_COVER_PERIOD, "Cover: Cover period is too long");
+    require(buyCoverParams.commissionRatio <= MAX_COMMISSION_RATIO, "Cover: Commission rate is too high");
+
+
+    // Override cover specific parameters
+    buyCoverParams.payoutAsset = cover.payoutAsset;
+    buyCoverParams.productId = cover.productId;
+
+    uint refundInCoverAsset = 0;
+
+    if (!lastCoverSegment.expired) {
+      uint32 remainingPeriod = lastCoverSegment.start + lastCoverSegment.period - uint32(block.timestamp);
+
+      PoolAllocation[] storage originalPoolAllocations = coverSegmentAllocations[coverId][lastCoverSegmentIndex];
+
+      {
+        uint originalPoolAllocationsCount = originalPoolAllocations.length;
+
+        // rollback previous cover
+        for (uint i = 0; i < originalPoolAllocationsCount; i++) {
+          stakingPool(originalPoolAllocations[i].poolId).deallocateStake(
+            cover.productId,
+            lastCoverSegment.start,
+            lastCoverSegment.period,
+            originalPoolAllocations[i].coverAmountInNXM,
+            originalPoolAllocations[i].premiumInNXM,
+            lastCoverSegment.globalRewardsRatio
+          );
+
+          originalPoolAllocations[i].premiumInNXM *= (lastCoverSegment.period - remainingPeriod) / lastCoverSegment.period;
+
+          // compute NXM rewards deallocated
+          uint deallocatedRewardsInNXM = originalPoolAllocations[i].premiumInNXM
```

Comment:
It is likely that this won't be *exactly* directly proportional. Please add a todo to use a value returned by `deallocateStake`